### ISSUE TITLE
(Feature): Natural voice playback using tts for edge

### ISSUE THREAD

**Issue by Iheuzio** [Osmodium/PathfinderTextToSpeechMod#25]
Hi,

I've recently saw the project here and it is using the services directly in windows. However the natural voices do sound much better, but are locked from the normal api in windows. They could be accessed with the edge/chrome api through `chrome.ttsEngine` or `chrome.tts`. 

The added benefit is you do not need windows to run this and you can relay the calls to windows, or possibly whatever text reader is configured for the browser. There is the github project [here](https://github.com/rany2/edge-tts/), the work is already done there and you can simply pass the command from unity into the edge-tts command from the program there. 

If this sounds reasonable, I could try working on a possible integration.

**Comment by Osmodium** [comment]
Hi!
Thanks for your interest. 
There are many who have suggested using some other library/framework for the speech generation (like [Bark](https://github.com/suno-ai/bark)).
You are correct that it uses Windows directly (if on windows) and it uses the OSX build-in tts feature when on OSX.

I will look into edge-tts, since it has a feature to have the audio play back immediately.
Thanks for your suggestion.

**Comment by Osmodium** [comment]
I'm just realizing; are you talking about these voices? https://www.ghacks.net/2018/08/11/unlock-all-windows-10-tts-voices-system-wide-to-get-more-of-them/
Because then the mod already supports the voices, you "just" have to enable them through the guide.

**Comment by Iheuzio** [comment]
It should read whichever tts voice is configured in chrome/chromium. This is separate from windows' voices and would require to pass the text through the browser API in order to be read.

**Comment by Osmodium** [comment]
So this would require the user to have chrome installed?

**Comment by Iheuzio** [comment]
https://github.com/Iheuzio/PathfinderTextToSpeechMod

it requires ffmpeg, mpv and edge-tts to be installed. Then it would work like this:
https://cdn.discordapp.com/attachments/791053285657542666/1177332422556262480/Untitled_video_-_Made_with_Clipchamp.mp4?ex=65721f43&is=655faa43&hm=c862fd6ab738179d58ee494d3bb4a65382663212ffb49108bfd12827fc03e117&

**Comment by Iheuzio** [comment]
If you want to communicate on discord I'm in the server `@iheuzio`.

I could integrate a change where there is an option in the menu that allows this option to be toggled on or off. As well as a script to setup the mpv, ffmpeg and all that on windows to make it easier for installation. Let me know if that sounds good.

**Comment by Osmodium** [comment]
Thanks for providing a PoC of it, and no worries about taking time :)
I haven't had time to look at the fork yet but, it looks a sounds good.
I have some requests/concers:
It sounds like it takes a while for the audio to play after clicking, can this be reduced?
I imagine needing to have some checks as to if the computer has the applications installed and the correct supported versions of them, have you included this?

**Comment by Iheuzio** [comment]
- It sounds like it takes a while for the audio to play after clicking, can this be reduced?

 > Yes, the longer the text, the longer it takes to save a temporary mp3 file, we could either split up the mp3 files into many smaller ones and play them separately as the others finish up sequentially. 

- I imagine needing to have some checks as to if the computer has the applications installed and the correct supported versions of them, have you included this?

> I did not test that, this was done just to show that using edge-tts was possible. I can work on an actual integration later, however my code was written in 30 minutes so it's pretty bad.

**Comment by adamstradomski** [comment]
Is this idea stopped? I would love to see it work. Especially with Rouge Trader :) 


But I believe the edge-tts does require internet connection. It's a python wrapper over Bing API? 
![image](https://github.com/Osmodium/PathfinderTextToSpeechMod/assets/83371226/229fdd85-bf89-4527-af21-961e45f197a3)

**Comment by Iheuzio** [comment]
Yeah, it requires internet to be connected for natural voices. My bad, however I'll be able to work on a simple proper integration with the existing tts possibly after january 12th.

**Comment by Osmodium** [comment]
@Iheuzio Saw your video on the owl-cat forum. Looks promising with a bit of delay still 👍

**Comment by Osmodium** [comment]
I just dabbled a bit with this in LinqPad, and I got it to work without having the python program installed, which would probably be preferred?

**Comment by Osmodium** [comment]
Hmm, it seems that I can get it to stream, so I dont have to wait for the entire thing to be sent from the server, which cuts significantly down on the load/latency times.
However, this is only if I use +0hz pitch, +0% rate and +0% volume. If I change any of them, it increases the response time by a huge margin (from 0.5 sec to 6.5 sec).

**Comment by Wazard** [comment]
hey, are you still working on that idea? I do have experience in coding wiht C# and unity but none in tts so i don't know if I would be of any help

**Comment by Osmodium** [comment]
@Wazard Yes, and I've gotten it to work, but discovered that the service that is being used to generate the audio does not support multiple voices in one request. So I'm currently working on parallelizing calls for each section of the dialog.

**Comment by Wazard** [comment]
@Osmodium sorry for the probably dumb question, but: I saw that from the narrator you can add and download the natural voices. Couldn't be possible to use em within windows itself in this way?

**Comment by Iheuzio** [comment]
Hi, currently I won't have time to focus on this mod. I'm working on other stuff at the moment, I may be able to try helping out later however not for the meantime.

**Comment by bubval** [comment]
Hey @Osmodium it's awesome ot hear that you have it working. Would it be possible to have a release not including multiple voices? How's the progress so far?

**Comment by Osmodium** [comment]
Hi! I have just uploaded the experimental version of the mod (0.9.4-EXP) [here](https://www.nexusmods.com/warhammer40kroguetrader/mods/75?tab=files) which includes Natural Voices through the Bing service. It is the version from over a month ago, but progress has been slow.
It is still WIP so there might be bugs.

**Comment by Christian-Arning** [comment]
@Osmodium does this mod also work for pathfinder wotr, if not how can i make it work there?

**Comment by Osmodium** [comment]
@BelegCufea That looks awesome and interesting! It looks like it only works for Windows 11, or I might be missing something.

**Comment by BelegCufea** [comment]
@Osmodium I have no idea, but on the Repo page there is a mention about Win 10 in `System Requirements` section:


**I'm using Windows 10. Can I use the Narrator natural voices on Windows 11?**

Yes, as long as your Windows 10 build number is 17763 or above (version 1809). You can choose and install Windows 11 Narrator voices [here](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter/wiki/Narrator-natural-voice-download-links).

Windows 10's Narrator doesn't support natural voices directly, but it does support SAPI 5 voices. So you can make Windows 11 Narrator voices work on Windows 10 via this engine.

**Comment by Osmodium** [comment]
@BelegCufea I will test this out, since I'm on a Windows 10. Thanks!

**Comment by Osmodium** [comment]
It seems to be working pretty well, apart from it crashing when issued to stop. I have to look into this, but otherwise this is a pretty elegant solution for those who want those voices, and possibly others too.

**Comment by BelegCufea** [comment]
@Osmodium Nice. Works fine on Win 11. No crashing at all. Even when interrupting dialogs.

Sometimes there is a slight delay when using Edge voices, but not always. And even when there is one, it is acceptable (me thinks :-) )

And it seems it has some problem with `<silence/>` tag. Had to change phonetics like so:

```
  "—": " . . ",
  "...": " . . . ", 
```

**Comment by BelegCufea** [comment]
@Osmodium Update after a few more hours of play.

- Occasionally (about 1% of the time), the game becomes unresponsive when starting a new "page" of dialogue.
- I always wait for the dialogue to be fully read before proceeding, so there is no interrupting.
- This only happens when using Edge voices.
- I can't determine where the freeze is occurring: in SAPIAdapter, in the WindowsVoice DLL, or in the C# wrapper.
- As far as I can tell, this has never happened to me with offline Natural voices.

**Comment by LeapSoftware** [comment]
Hey All, just thought id add my own experience to the above mentioned ^. 

I cloned the repo and made the changes you mentioned above (plus using NaturlaVoiceSAPIAdapter). I was able to get it working and detecting all online and locally downloaded natural voices. After using for a bit i can confirm that using online voices seems to every so often hang the game indefinitely. I have not delved into where exactly it is throwing (i cant see any errors) but there is definitely an issue there.

If I find any other issues ill post here :)

**Comment by BelegCufea** [comment]
@LeapSoftware Thanks for info.

Unfortunately, that is true. It is unstable when using online voices. Nevertheless, I have had no problems so far using offline natural voices.

For anyone interested, I compiled the changes at [my fork](https://github.com/BelegCufea/PathfinderTextToSpeechMod/releases/tag/1.0.9-NV). It is highly experimental though! Use at your own risk :-)

**Comment by Wazard** [comment]
> @LeapSoftware Thanks for info.
> 
> Unfortunately, that is true. It is unstable when using online voices. Nevertheless, I have had no problems so far using offline natural voices.
> 
> For anyone interested, I compiled the changes at [my fork](https://github.com/BelegCufea/PathfinderTextToSpeechMod/releases/tag/1.0.9-NV). It is highly experimental though! Use at your own risk :-)

how do i make it to work? The only 2 available are Zira and David. I'm on win11 with natural voices installed

**Comment by Osmodium** [comment]
I'll add the code to the project and I guess I can do a small writeup about having to use [NaturalVoiceSAPIAdapter ](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter) to make it work.

**Comment by Christian-Arning** [comment]
> > @LeapSoftware Thanks for info.
> > Unfortunately, that is true. It is unstable when using online voices. Nevertheless, I have had no problems so far using offline natural voices.
> > For anyone interested, I compiled the changes at [my fork](https://github.com/BelegCufea/PathfinderTextToSpeechMod/releases/tag/1.0.9-NV). It is highly experimental though! Use at your own risk :-)
> 
> how do i make it to work? The only 2 available are Zira and David. I'm on win11 with natural voices installed

You have to install the [Natural Voice SAPI Adapter](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter) to make it show up.

**Comment by Philemon5** [comment]
I am getting the following error

[SpeechMod] [Exception] OnGUI: ArgumentException - An item with the same key has already been added. Key: Microsoft Neerja Online 
System.ArgumentException: An item with the same key has already been added. Key: Microsoft Neerja Online 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].TryInsert (TKey key, TValue value, System.Collections.Generic.InsertionBehavior behavior) [0x000c1] in <695d1cc93cca45069c528c15c9fdd749>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Add (TKey key, TValue value) [0x00000] in <695d1cc93cca45069c528c15c9fdd749>:0 
  at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector, System.Collections.Generic.IEqualityComparer`1[T] comparer) [0x0009d] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement] (System.Collections.Generic.IEnumerable`1[T] source, System.Func`2[T,TResult] keySelector, System.Func`2[T,TResult] elementSelector) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at SpeechMod.Main.get_VoicesDict () [0x0003a] in <abffca3fc1554b15b58c0755c7cba401>:0 
  at SpeechMod.Unity.MenuGUI.AddVoiceSelector (System.String label, System.Int32& voice, System.String& previewString, System.Int32& rate, System.Int32& volume, System.Int32& pitch, SpeechMod.Voice.VoiceType type) [0x00047] in <abffca3fc1554b15b58c0755c7cba401>:0 
  at SpeechMod.Unity.MenuGUI.OnGui () [0x00000] in <abffca3fc1554b15b58c0755c7cba401>:0 
  at SpeechMod.Main.OnGui (UnityModManagerNet.UnityModManager+ModEntry modEntry) [0x00007] in <abffca3fc1554b15b58c0755c7cba401>:0 
  at UnityModManagerNet.UnityModManager+UI.DrawTab (System.Int32 tabId, UnityEngine.Events.UnityAction& buttons) [0x007d2] in <e2b836aad626492cb17d508e50751848>:0

**Comment by Philemon5** [comment]
in any case the fork provided doesnt seem to work for me

**Comment by BelegCufea** [comment]
@Philemon5 It is only tested for English (US+UK). If it is sufficient for you, just run NaturalVoiceSAPIAdapter Installer again and change Included languages to either "Follow user's preferred languages" or pick English(US) + English(UK). "Microsoft Neerja" is in English(India) so definitely don't pick that one! BTW no need to reinstall SAPIAdapter.

![image](https://github.com/user-attachments/assets/62cc7e52-bf71-4259-8b13-a6f24395531a)

**Comment by Philemon5** [comment]
okay but every voice only plays once in preview and then nothing. and it doesnt show the voices i downloaded from microsoft store

**Comment by Philemon5** [comment]
update: i installed windows 11 but is till dont see the offline voices in my system o.o

**Comment by BelegCufea** [comment]
I have a non-English version of Windows 11, so my screenshots might look like gibberish. However, I found a helpful walkthrough:

[How to Use Narrator’s New Natural Voices](https://www.partitionwizard.com/clone-disk/narrators-new-natural-voices-in-windows-11.html)

especially: `Step 4: You will go back to the Narrator Settings window, and you need to click the Add button under the Add natural voices section.`

BTW the online voices are bit finicky in Pathfinder, better to use these offline ones.

**Comment by Philemon5** [comment]
Sadly the option to add natural voices does not show up on my windows 11 install, and i have tried removing the old voices i added in windows 10, with no luck. I tried uinstalling the old voice packages with consle commands and restarting and they still show up. 
![image](https://github.com/user-attachments/assets/39c8850c-6479-4999-8daf-3a780fef5bdf)

**Comment by Philemon5** [comment]
It does work now!

**Comment by jarrard** [comment]
Yeah I got this working fine.  
It seems to me the crashing happens during or soon after you do a lot of tweaking in the settings of the mod.
Whereas if you just leave it alone it works for much longer.

**Comment by Osmodium** [comment]
I have updated the mod to support Natrural Voices based on this issue, if you want to go back to following the releases :)
Anywho I'm going to close this issue soon, unless there's more to discuss here.

**Comment by jarrard** [comment]
I hope you can figure out what is causing this freeze/lockup issue. Seems to happen after a little while, kind of random.

**Comment by Osmodium** [comment]
@jarrard are you using the online voices or the locally installed ones?

**Comment by jarrard** [comment]
Online

**Comment by Osmodium** [comment]
Yeah. there is a known issue with the online voices, that I have narrowed down to an issue in the [NaturalSAPIAdapter](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter/wiki/Narrator-natural-voice-download-links) which crashes when trying to stop and release the current playback. I hope when they get further in the implementation that the issue will be resolved.

**Comment by jarrard** [comment]
Yeah I downloaded them but it AUTOINSTALLS into a unknown folder, I can't tell it where they went.  Instructions seem outdated as they talk about unzipping it and directing the API there, no luck.

**Comment by gexgd0419** [comment]
Hi @Osmodium , I've released a new version [v0.2.1](https://github.com/gexgd0419/NaturalVoiceSAPIAdapter/releases/tag/v0.2.1) of NaturalSAPIAdapter.

I've changed a lot of code which handles online voices, so I want to know if it becomes better or worse.

Unfortunately I don't have this game, so I cannot test this myself.

Could someone test whether the online voices work well using NaturalVoiceSAPIAdapter v0.2.1?

**Comment by BelegCufea** [comment]
Hi @gexgd0419 .

I am not playing actively atm, but I have loaded some of my saves and tried talking to many NPCs for 15 minutes or so.
I switched all voices to use Online (Edge) variants.

Everything seems fine now. I have not encountered a single deadlock. There is slight delay at the beginning of every dialog (as to be expected) but it is berly noticeable (250-500ms). Switching from one voice to another in the same dialog is seamless (no delay at all - great work there).

Maybe someone else can report more while actively playing. Especially regarding deadlocking the game while using Edge voices (as this was the main problem).

Thanks for your awesome work!

FYI. I am using your adapter in WoW fot TTS as well and there is no problem at all.
